src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -902,6 +902,25 @@ public SolutionState WithProjectReferences(ProjectId projectId, IEnumerable<Proj
             return this.ForkProject(newProject, newDependencyGraph: _dependencyGraph.WithProjectReferences(projectId, projectReferences.Select(p => p.ProjectId)));
         }
 
+        /// <summary>
+        /// Creates a new solution instance with the project documents in the order by the specified document ids.
+        /// The specified document ids must exist within the specified project and have the same count as the number of documents in the specified project.
+        /// </summary>
+        public SolutionState WithProjectDocumentsOrder(ProjectId projectId, ImmutableArray<DocumentId> documentIds)
+        {
+            if (projectId == null)
+            {
+                throw new ArgumentNullException(nameof(projectId));
+            }
+
+            CheckContainsProject(projectId);
+
+            var oldProject = this.GetProjectState(projectId);
+            var newProject = oldProject.UpdateDocumentsOrder(documentIds);
+
+            return this.ForkProject(newProject);
```

Comment:
What happens in the C#/VB case? Shouldn't this be reordering the syntax trees?

Reply:
Good question because I don't know how the syntax trees have to do with it.

Reply:
Yeah, we need to ensure in the C#/VB case that the compilation's file order matches the new thing. This means we probably need to stick a translation action into here.